onnxruntime/core/framework/tensorprotoutils.cc

```diff
@@ -114,10 +114,114 @@ static Status UnpackTensorWithRawData(const void* raw_data, size_t raw_data_leng
       gsl::make_span(raw_data_bytes, raw_data_length), gsl::make_span(p_data, expected_size)));
   return Status::OK();
 }
+
+static Status GetExternalDataInfo(const ONNX_NAMESPACE::TensorProto& tensor_proto,
+                                  const ORTCHAR_T* tensor_proto_dir,
+                                  std::basic_string<ORTCHAR_T>& external_file_path,
+                                  onnxruntime::FileOffsetType& file_offset,
+                                  size_t& tensor_data_length) {
+  ORT_RETURN_IF(tensor_proto.data_location() != TensorProto_DataLocation_EXTERNAL,
+                "Tensor does not have external data to read from.");
+
+  ORT_RETURN_IF_NOT(
+      tensor_proto.data_type() != ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED &&
+          tensor_proto.data_type() != ONNX_NAMESPACE::TensorProto_DataType_STRING,
+      "External data type cannot be UNDEFINED or STRING.");
+
+  std::unique_ptr<onnxruntime::ExternalDataInfo> external_data_info;
+  ORT_RETURN_IF_ERROR(onnxruntime::ExternalDataInfo::Create(tensor_proto.external_data(), external_data_info));
+
+  if (tensor_proto_dir != nullptr) {
+    external_file_path = onnxruntime::ConcatPathComponent<ORTCHAR_T>(tensor_proto_dir, external_data_info->GetRelPath());
+  } else {
+    external_file_path = external_data_info->GetRelPath();
+  }
+
+  file_offset = external_data_info->GetOffset();
+
+  ORT_RETURN_IF_ERROR(onnxruntime::utils::GetSizeInBytesFromTensorProto<0>(
+      tensor_proto, &tensor_data_length));
+  const size_t external_data_length = external_data_info->GetLength();
+
+  ORT_RETURN_IF_NOT(
+      external_data_length == 0 ||
+          external_data_length == tensor_data_length,
+      "TensorProto external data size mismatch. ",
+      "Computed size: ", tensor_data_length,
+      ", external_data.length: ", external_data_length);
+
+  return Status::OK();
+}
+
+// Read external data for tensor in unint8_t* form and return Status::OK() if the data is read successfully.
+// Uses the tensor_proto_dir to construct the full path for external data. If tensor_proto_dir == nullptr
+// then uses the current directory instead.
+// This function does not unpack string_data of an initializer tensor
+static Status ReadExternalDataForTensor(const ONNX_NAMESPACE::TensorProto& tensor_proto,
+                                        const ORTCHAR_T* tensor_proto_dir,
+                                        std::unique_ptr<uint8_t[]>& unpacked_tensor,
+                                        size_t& tensor_data_length) {
```

Comment:
No, you are supposed to use things like https://github.com/microsoft/onnxruntime/blob/eb5c1f0fcc6038f2743e7f5d2c1c43a014bdfdcb/onnxruntime/core/framework/tensorprotoutils.cc#L388

Please don't call ReadFileIntoBuffer directly.

Reply:
You may move the GetFileContent function to the Env class if needed.

Reply:
Marking it as resolved per offline discussion. In this case it is OK to use ReadFileIntoBuffer as we load the data in memory managed by unique_ptr...